File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/library/LibraryImportedSourcesStore.kt
package com.pillyliu.pinprofandroid.library

import android.content.Context
import androidx.core.content.edit
import org.json.JSONArray
import org.json.JSONObject

internal object ImportedSourcesStore {
    private const val PREFS_NAME = "practice-upgrade-state-v2"
    private const val SOURCES_KEY = "pinball-imported-sources-v1"

    fun hasPersistedSources(context: Context): Boolean {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        return prefs.contains(SOURCES_KEY)
    }

    fun load(context: Context): List<ImportedSourceRecord> {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val raw = prefs.getString(SOURCES_KEY, null)
            ?: return if (!LibrarySourceStateStore.hasPersistedState(context)) {
                bundledDefaultImportedSources().also { defaults ->
                    if (defaults.isNotEmpty()) {
                        save(context, defaults)
                    }
                }
            } else {
                emptyList()
            }
        return runCatching {
            val array = JSONArray(raw)
            val records = buildList {
                for (i in 0 until array.length()) {
                    val obj = array.optJSONObject(i) ?: continue
                    val id = canonicalLibrarySourceId(obj.optString("id").trim())
                    val name = obj.optString("name").trim()
                    val type = LibrarySourceType.fromRaw(obj.optString("type")) ?: continue
                    val provider = ImportedSourceProvider.fromRaw(obj.optString("provider"))
                        ?: inferredImportedSourceProvider(type, id)
                    val providerSourceId = obj.optString("providerSourceId").trim()
                    if (id.isNullOrBlank() || name.isBlank() || providerSourceId.isBlank()) continue
                    add(
                        ImportedSourceRecord(
                            id = id,
                            name = name,
                            type = type,
                            provider = provider,
                            providerSourceId = providerSourceId,
                            machineIds = obj.optJSONArray("machineIds").toStringList(),
                            lastSyncedAtMs = obj.optLong("lastSyncedAtMs").takeIf { it > 0L },
                            searchQuery = obj.optString("searchQuery").trim().ifBlank { null },
                            distanceMiles = obj.optInt("distanceMiles").takeIf { it > 0 },
                        ),
                    )
                }
            }
            val migrated = normalizedImportedRecords(records)
            if (migrated != records) {
                save(context, migrated)
            }
            migrated
        }.getOrDefault(emptyList())
    }

    fun save(context: Context, records: List<ImportedSourceRecord>) {
        val normalized = normalizedImportedRecords(records)
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val json = JSONArray().apply {
            normalized.forEach { record ->
                put(
                    JSONObject().apply {
                        put("id", record.id)
                        put("name", record.name)
                        put("type", record.type.rawValue)
                        put("provider", record.provider.rawValue)
                        put("providerSourceId", record.providerSourceId)
                        put("machineIds", JSONArray().apply { record.machineIds.forEach(::put) })
                        put("lastSyncedAtMs", record.lastSyncedAtMs)
                        put("searchQuery", record.searchQuery)
                        put("distanceMiles", record.distanceMiles)
                    },
                )
            }
        }
        prefs.edit { putString(SOURCES_KEY, json.toString()) }
    }

    fun upsert(context: Context, record: ImportedSourceRecord) {
        val canonicalRecord = normalizedImportedRecord(record) ?: return
        val canonicalId = canonicalRecord.id
        val current = load(context).toMutableList()
        val index = current.indexOfFirst { it.id == canonicalId }
        if (index >= 0) {
            current[index] = canonicalRecord
        } else {
            current += canonicalRecord
        }
        save(context, current)
    }

    fun remove(context: Context, id: String) {
        val canonicalId = canonicalLibrarySourceId(id) ?: return
        save(context, load(context).filterNot { it.id == canonicalId })
        LibrarySourceStateStore.removeSourcePreferences(context, canonicalId)
    }
}

File: Pinball App 2/Pinball App 2/library/LibraryImportedSourcesStore.swift
import Foundation

enum PinballImportedSourcesStore {
    private static let defaultsKey = "pinball-imported-sources-v1"

    static func hasPersistedSources() -> Bool {
        UserDefaults.standard.data(forKey: defaultsKey) != nil
    }

    static func load() -> [PinballImportedSourceRecord] {
        guard let data = UserDefaults.standard.data(forKey: defaultsKey) else {
            guard !PinballLibrarySourceStateStore.hasPersistedState() else {
                return []
            }
            let defaults = loadBundledDefaults()
            if !defaults.isEmpty {
                save(defaults)
            }
            return defaults
        }
        if let records = try? JSONDecoder().decode([PinballImportedSourceRecord].self, from: data) {
            let migrated = normalizedImportedRecords(records)
            if migrated != records {
                save(migrated)
            }
            return migrated
        }
        guard let array = (try? JSONSerialization.jsonObject(with: data)) as? [[String: Any]] else {
            return []
        }
        let iso = ISO8601DateFormatter()
        let records = array.compactMap { item -> PinballImportedSourceRecord? in
            guard let id = canonicalLibrarySourceID(item["id"] as? String),
                  let name = (item["name"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines), !name.isEmpty,
                  let typeRaw = item["type"] as? String,
                  let type = PinballLibrarySourceType(rawValue: typeRaw),
                  let providerSourceID = (item["providerSourceID"] as? String ?? item["providerSourceId"] as? String)?
                    .trimmingCharacters(in: .whitespacesAndNewlines),
                  !providerSourceID.isEmpty else {
                return nil
            }
            let provider = (item["provider"] as? String).flatMap(PinballImportedSourceProvider.init(rawValue:))
                ?? inferredImportedSourceProvider(type: type, id: id)
            let lastSyncedAt = (item["lastSyncedAt"] as? String).flatMap { iso.date(from: $0) }
            let machineIDs = (item["machineIDs"] as? [Any] ?? item["machineIds"] as? [Any] ?? []).compactMap {
                ($0 as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
            }.filter { !$0.isEmpty }
            return PinballImportedSourceRecord(
                id: id,
                name: name,
                type: type,
                provider: provider,
                providerSourceID: providerSourceID,
                machineIDs: machineIDs,
                lastSyncedAt: lastSyncedAt,
                searchQuery: (item["searchQuery"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines),
                distanceMiles: item["distanceMiles"] as? Int
            )
        }
        let migrated = normalizedImportedRecords(records)
        save(migrated)
        return migrated
    }

    static func save(_ records: [PinballImportedSourceRecord]) {
        let normalized = normalizedImportedRecords(records)
        guard let data = try? JSONEncoder().encode(normalized) else { return }
        UserDefaults.standard.set(data, forKey: defaultsKey)
    }

    static func upsert(_ record: PinballImportedSourceRecord) {
        var records = load()
        guard let canonicalRecord = normalizedImportedRecord(record) else { return }
        if let index = records.firstIndex(where: { $0.id == canonicalRecord.id }) {
            records[index] = canonicalRecord
        } else {
            records.append(canonicalRecord)
        }
        save(records)
    }

    static func remove(id: String) {
        guard let id = canonicalLibrarySourceID(id) else { return }
        var records = load()
        records.removeAll { $0.id == id }
        save(records)
        PinballLibrarySourceStateStore.removeSourcePreferences(sourceID: id)
    }
}
